consensus/src/pos/target.rs

```diff
@@ -0,0 +1,575 @@
+// Copyright (c) 2023 RBB S.r.l
+// opensource@mintlayer.org
+// SPDX-License-Identifier: MIT
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://github.com/mintlayer/mintlayer-core/blob/master/LICENSE
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use chainstate_types::{BlockIndex, BlockIndexHandle};
+use common::{
+    chain::{
+        block::ConsensusData, ChainConfig, GenBlock, GenBlockId, PoSChainConfig, PoSStatus,
+        RequiredConsensus,
+    },
+    primitives::{BlockDistance, BlockHeight, Compact, Id},
+    Uint256, Uint512,
+};
+
+use crate::pos::error::ConsensusPoSError;
+
+fn calculate_average_block_time(
+    chain_config: &ChainConfig,
+    pos_config: &PoSChainConfig,
+    block_index: &BlockIndex,
+    block_index_handle: &impl BlockIndexHandle,
+) -> Result<u64, ConsensusPoSError> {
+    // Determine net upgrade version that current block belongs to.
+    // Across versions config parameters might change so it's invalid to mix versions for target calculations
+    let (_, net_version) = chain_config
+        .net_upgrade()
+        .version_at_height(block_index.block_height())
+        .expect("NetUpgrade must've been initialized");
+    let net_version_range = chain_config
+        .net_upgrade()
+        .height_range(net_version)
+        .expect("NetUpgrade must've been initialized");
+
+    // Average is calculated based on 2 timestamps and then is divided by number of blocks in between.
+    // Choose a block from the history that would be the start of a timespan.
+    // It shouldn't cross net version range or genesis.
+    let timespan_start_height = std::cmp::max(
+        net_version_range.start,
+        (block_index.block_height()
+            - BlockDistance::new(pos_config.block_count_to_average_for_blocktime() as i64))
+        .unwrap_or(BlockHeight::zero()),
+    );
+
+    let time_span_start = block_index_handle
+        .get_ancestor(block_index, timespan_start_height)?
+        .block_timestamp()
+        .as_int_seconds();
+    let current_block_time = block_index.block_timestamp().as_int_seconds();
+
+    let blocks_in_timespan: i64 = (block_index.block_height() - timespan_start_height)
+        .expect("cannot be negative")
+        .into();
+
+    let average = (current_block_time - time_span_start) / blocks_in_timespan as u64;
+
+    Ok(average)
+}
+
+fn calculate_new_target(
+    pos_config: &PoSChainConfig,
+    prev_target: &Uint256,
+    average_block_time: u64,
```

Comment:
I think this should be called `block_time`... it's an assumption that it's an "average". It depends on other function calls.